v2/tools/generator/internal/codegen/code_generator.go

```diff
@@ -225,29 +226,25 @@ func (generator *CodeGenerator) Generate(ctx context.Context) error {
 	state := pipeline.NewState()
 	for i, stage := range generator.pipeline {
 		klog.V(0).Infof("%d/%d: %s", i+1, len(generator.pipeline), stage.Description())
-		// Defensive copy (in case the pipeline modifies its inputs) so that we can compare types in vs out
-		stateOut, err := stage.Run(ctx, state)
+
+		start := time.Now()
+
+		newState, err := stage.Run(ctx, state)
 		if err != nil {
 			return errors.Wrapf(err, "failed during pipeline stage %d/%d: %s", i+1, len(generator.pipeline), stage.Description())
 		}
 
 		// Fail fast if something goes awry
-		if len(stateOut.Definitions()) == 0 {
+		if len(newState.Definitions()) == 0 {
 			return errors.Errorf("all type definitions removed by stage %s", stage.Id())
 		}
 
-		defsAdded := stateOut.Definitions().Except(state.Definitions())
-		defsRemoved := state.Definitions().Except(stateOut.Definitions())
+		generator.logStateChange(state, newState)
 
-		if len(defsAdded) > 0 && len(defsRemoved) > 0 {
-			klog.V(1).Infof("Added %d, removed %d type definitions", len(defsAdded), len(defsRemoved))
-		} else if len(defsAdded) > 0 {
-			klog.V(1).Infof("Added %d type definitions", len(defsAdded))
-		} else if len(defsRemoved) > 0 {
-			klog.V(1).Infof("Removed %d type definitions", len(defsRemoved))
-		}
+		duration := time.Since(start).Round(time.Millisecond)
+		klog.V(0).Infof("%d/%d: completed in %s", i+1, len(generator.pipeline), duration)
```

Comment:
minor: I think this would possibly be easier to read if the pre and post logs had the same prefix.

Wondering if it makes sense to log: `klog.V(0).Infof("%d/%d: %s, completed in %s", i+1, len(generator.pipeline), stage.Description(), duration)` instead to match?

Reply:
Done, logs now look like this:

```
I0518 07:49:30.818128   30472 code_generator.go:228] 49/63: Mark the latest GA storage variant of each resource as the hub version
I0518 07:49:30.926754   30472 code_generator.go:249] 49/63: Mark the latest GA storage variant of each resource as the hub version, completed in 107ms
I0518 07:49:30.926754   30472 code_generator.go:228] 50/63: Inject the function Hub() into each hub resource
I0518 07:49:30.972481   30472 code_generator.go:249] 50/63: Inject the function Hub() into each hub resource, completed in 46ms
I0518 07:49:30.976483   30472 code_generator.go:228] 51/63: Implement the Convertible interface on each non-hub Resource type
I0518 07:49:31.026228   30472 code_generator.go:249] 51/63: Implement the Convertible interface on each non-hub Resource type, completed in 50ms
I0518 07:49:31.026228   30472 code_generator.go:228] 52/63: Add test cases to verify JSON serialization
I0518 07:49:31.114907   30472 code_generator.go:249] 52/63: Add test cases to verify JSON serialization, completed in 89ms
I0518 07:49:31.114907   30472 code_generator.go:228] 53/63: Add test cases to verify PropertyAssignment functions
I0518 07:49:31.192027   30472 code_generator.go:249] 53/63: Add test cases to verify PropertyAssignment functions, completed in 77ms
```